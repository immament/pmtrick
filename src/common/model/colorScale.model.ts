export class ColorScale {
    private delta: number;
    private lightnessDelta: number;
    constructor(
        public readonly maxValue: number,
        public readonly minValue = 0,
        public readonly hue = 255,
        public readonly saturation = 100,
        public readonly maxLightness = 90,
        public readonly minLightness = 25,
    ) {
        this.delta = this.maxValue - this.minValue;
        this.lightnessDelta = this.maxLightness - this.minLightness;
    }

    // lightness between 25% and 90%
    getHslColor(value: number): string {
        const normalizedValue = Math.min(Math.max(0, value - this.minValue), this.maxValue);
        return `hsl(${this.hue}, ${this.saturation}%, ${
            this.maxLightness - (normalizedValue / this.delta) * this.lightnessDelta
        }%`;
    }
}
