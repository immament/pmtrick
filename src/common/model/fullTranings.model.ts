export class FullTrainings {
    constructor({
        currentSeasonLeagueMatches,
        currentSeasonFriendlyMatches,
        fullSeasonLeagueMatches = 18,
        fullSeasonFriendlyMatches = 12,
    }: {
        currentSeasonLeagueMatches?: number;
        currentSeasonFriendlyMatches?: number;
        fullSeasonLeagueMatches?: number;
        fullSeasonFriendlyMatches?: number;
    }) {
        this.curentSeason = (currentSeasonLeagueMatches ?? 0) + (currentSeasonFriendlyMatches ?? 0) / 2;
        this.fullSeason = fullSeasonLeagueMatches + fullSeasonFriendlyMatches / 2;
    }

    curentSeason: number;
    fullSeason: number;
}
