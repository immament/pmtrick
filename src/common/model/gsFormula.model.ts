export interface GsFormula {
    main: number;
    secondary: number;
    tertiary: number;
    physical: number;
}

export const mainGsFormula: GsFormula = {
    main: 61.02,
    secondary: 20.34,
    physical: 15.25,
    tertiary: 1.13,
};

// TODO
export const tempGsFormula: GsFormula = {
    main: 60,
    secondary: 20,
    physical: 20,
    tertiary: 0,
};
