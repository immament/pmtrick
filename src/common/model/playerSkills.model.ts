export interface PlayerSkills {
    tackling: number;
    heading: number;
    passing: number;
    positioning: number;
    finishing: number;
    technique: number;
    speed: number;
    strength: number;
    handling: number;
    outOfArea: number;
    reflexes: number;
    agility: number;
}

export type PlayerSkillsType =
    | 'tackling'
    | 'heading'
    | 'passing'
    | 'positioning'
    | 'finishing'
    | 'technique'
    | 'speed'
    | 'strength'
    | 'handling'
    | 'outOfArea'
    | 'reflexes'
    | 'agility';

export enum PlayerSkillsEnum {
    'tackling' = 'tackling',
    'heading' = 'heading',
    'passing' = 'passing',
    'positioning' = 'positioning',
    'finishing' = 'finishing',
    'technique' = 'technique',
    'speed' = 'speed',
    'strength' = 'strength',
    'handling' = 'handling',
    'outOfArea' = 'outOfArea',
    'reflexes' = 'reflexes',
    'agility' = 'agility',
}
