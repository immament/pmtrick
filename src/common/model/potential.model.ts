export interface PotentialConfig {
    min: number;
    max: number;
}

const potentialConfig: PotentialConfig[] = [
    { min: 0.004, max: 0.048 }, // Terrible
    { min: 0.052, max: 0.096 }, // Very Bad
    { min: 0.1, max: 0.144 }, // Bad
    { min: 0.148, max: 0.192 }, // Low
    { min: 0.196, max: 0.24 }, // Passable
    { min: 0.244, max: 0.288 }, // Good
    { min: 0.292, max: 0.336 }, // Very Good
    { min: 0.34, max: 0.384 }, // Excellent
    { min: 0.388, max: 0.432 }, // Formidable
    { min: 0.436, max: 0.48 }, // World Class
];

export function getPotentialConfig(): PotentialConfig[] {
    return potentialConfig;
}
