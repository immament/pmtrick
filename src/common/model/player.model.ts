import { MatchesInSeasons } from '@src/common/services/storage/skillsTableOptions.repository';

import { SkillsSummary } from '../../modules/tacticSummary/model/skillsSummary.model';

import { PlayerSkills } from './playerSkills.model';

export type PlayerPositionType = 'G' | 'D' | 'M' | 'F';
export const playerPositionValues: PlayerPositionType[] = ['G', 'D', 'M', 'F'];

export enum PlayerPositionEnum {
    G,
    D,
    M,
    F,
}

export function isPlayerPosition(pos: string): pos is PlayerPositionType {
    return pos in playerPositionValues;
}

export enum PlayerQuality {
    'Terrible',
    'Very Bad',
    'Bad',
    'Low',
    'Passable',
    'Good',
    'Very Good',
    'Excellent',
    'Formidable',
    'World Class',
}

export type SkillsSummaries = {
    [position in PlayerPositionType]: SkillsSummary;
};

export interface PlayerWithSkillsSummaries extends Player {
    skillsSummaries?: SkillsSummaryCombo;
}

export interface SkillsSummaryWithBest extends SkillsSummaries {
    best?: SkillsSummary;
}

export interface FutureSkillsSummaryWithBest extends SkillsSummaryWithBest {
    options?: MatchesInSeasons;
}

export interface SkillsSummaryCombo {
    current?: SkillsSummaryWithBest;
    future?: FutureSkillsSummaryWithBest;
    futureMax?: FutureSkillsSummaryWithBest;
}

export interface Player {
    age: number;
    experience: PlayerQuality;
    experienceText: string;
    fitness: number;
    form?: string;
    id: number;
    maxTraining?: number;
    name: string;
    penalties: PlayerQuality;
    penaltiesText: string;
    position: string;
    positionZone: PlayerPositionType;
    potential: PlayerQuality;
    potentialText: string;
    quality: PlayerQuality;
    qualityText: string;
    skills: PlayerSkills;
    talent?: number;
    talentText?: string;
    temperament?: string | number; // TODO: enum
    value?: number;
    wage?: number;
}

export interface TlPlayer extends Player {
    deadline: string | number;
    price: string | number;
}
