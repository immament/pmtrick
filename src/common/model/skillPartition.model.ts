import { PlayerPositionType } from './player.model';
import { PlayerSkills } from './playerSkills.model';

type Skills = keyof PlayerSkills;

export enum SkillsGroup {
    main = 'main',
    secondary = 'secondary',
    tertiary = 'tertiary',
    // physical = 'physical'
}

export type SkillPartition = {
    [group in SkillsGroup]: Skills[];
};

export type SkillPartitions = {
    [position in PlayerPositionType]: SkillPartition;
};

const skillsPartitions: SkillPartitions = {
    G: {
        main: ['handling', 'outOfArea'],
        secondary: ['reflexes', 'agility'],
        tertiary: ['tackling', 'heading', 'passing', 'positioning', 'finishing', 'technique'],
    },
    D: {
        main: ['tackling', 'heading'],
        secondary: ['passing', 'positioning'],
        tertiary: ['handling', 'outOfArea', 'reflexes', 'agility', 'finishing', 'technique'],
    },
    M: {
        main: ['passing', 'positioning'],
        secondary: ['tackling', 'technique'],
        tertiary: ['handling', 'outOfArea', 'reflexes', 'agility', 'heading', 'finishing'],
    },
    F: {
        main: ['finishing', 'technique'],
        secondary: ['positioning', 'heading'],
        tertiary: ['handling', 'outOfArea', 'reflexes', 'agility', 'tackling', 'passing'],
    },
};

export function getSkillsPartitions(): SkillPartitions {
    return skillsPartitions;
}
