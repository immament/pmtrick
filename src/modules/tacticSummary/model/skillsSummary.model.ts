import { PlayerPositionType } from '../../../common/model/player.model';

export interface SkillsSummary {
    main: number;
    secondary: number;
    tertiary: number;
    physical: number;
    gs: number;
    position: PlayerPositionType;
    rank?: number;
}

type SkillsGroups = 'main' | 'secondary' | 'physical' | 'tertiary';

export const skillsSummaryOrder: SkillsGroups[] = ['main', 'secondary', 'physical', 'tertiary'];
