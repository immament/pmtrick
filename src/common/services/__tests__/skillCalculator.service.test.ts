import 'reflect-metadata';

import { mainGsFormula } from '@src/common/model/gsFormula.model';
import { Player, SkillsSummaries } from '@src/common/model/player.model';
import { SkillsSummary } from '@src/modules/tacticSummary/model/skillsSummary.model';

import { SkillCalculatorService } from '../skillCalculator.service';

describe('SkillCalculatorService', () => {
    it('Calculate gs for skills summary', () => {
        const service = new SkillCalculatorService(mainGsFormula);

        const skillsSummary: SkillsSummary = {
            main: 36,
            secondary: 22,
            physical: 22,
            tertiary: 31,
            position: 'G',
            gs: 0,
        };
        const result = service.gs(skillsSummary);
        expect(result).toBeCloseTo(74.7844, 4);
    });

    it('Calculate skills summaries for all positions', () => {
        const service = new SkillCalculatorService(mainGsFormula);
        const player: Partial<Player> = {
            skills: {
                handling: 16,
                outOfArea: 20,
                reflexes: 9,
                agility: 13,
                tackling: 7,
                heading: 5,
                passing: 7,
                positioning: 5,
                finishing: 2,
                technique: 5,
                speed: 14,
                strength: 8,
            },
            positionZone: 'G',
        };

        const skillsSummary = service.calculateGs(player as Player);

        expect(skillsSummary).toMatchSnapshot();
    });

    it('Should chould best skills summary', () => {
        const service = new SkillCalculatorService(mainGsFormula);

        const expectedBestSummary = {
            gs: 74.78441666666667,
            main: 36,
            physical: 22,
            position: 'G',
            rank: undefined,
            secondary: 22,
            tertiary: 31,
        };

        const summaries: SkillsSummaries = {
            D: {
                gs: 33.407583333333335,
                main: 12,
                physical: 22,
                position: 'D',
                rank: undefined,
                secondary: 12,
                tertiary: 65,
            },
            F: {
                gs: 24.829,
                main: 7,
                physical: 22,
                position: 'F',
                rank: undefined,
                secondary: 10,
                tertiary: 72,
            },
            G: {
                gs: 74.78441666666667,
                main: 36,
                physical: 22,
                position: 'G',
                rank: undefined,
                secondary: 22,
                tertiary: 31,
            },
            M: {
                gs: 33.407583333333335,
                main: 12,
                physical: 22,
                position: 'M',
                rank: undefined,
                secondary: 12,
                tertiary: 65,
            },
        };

        const bestSkillsSummary = service.chooseBestSummary(summaries);

        expect(bestSkillsSummary).toStrictEqual(expectedBestSummary);
        expect(bestSkillsSummary).toBe(summaries.G);
    });
});
