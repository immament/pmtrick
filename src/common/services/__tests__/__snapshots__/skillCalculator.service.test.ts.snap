// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillCalculatorService Calculate skills summaries for all positions 1`] = `
Object {
  "D": Object {
    "gs": 33.407583333333335,
    "main": 12,
    "physical": 22,
    "position": "D",
    "secondary": 12,
    "tertiary": 65,
  },
  "F": Object {
    "gs": 24.829,
    "main": 7,
    "physical": 22,
    "position": "F",
    "secondary": 10,
    "tertiary": 72,
  },
  "G": Object {
    "gs": 74.78441666666667,
    "main": 36,
    "physical": 22,
    "position": "G",
    "secondary": 22,
    "tertiary": 31,
  },
  "M": Object {
    "gs": 33.407583333333335,
    "main": 12,
    "physical": 22,
    "position": "M",
    "secondary": 12,
    "tertiary": 65,
  },
  "best": Object {
    "gs": 74.78441666666667,
    "main": 36,
    "physical": 22,
    "position": "G",
    "secondary": 22,
    "tertiary": 31,
  },
}
`;
